### PR TITLE
Calculate indices only when required.

### LINKED ISSUE #5379: ~70% Performance speedup for polygon_mesh_to_nef_3
## Issue Details

I have been investigating the performance of nef polyhedron construction, and have noticed the calculation of normals using normal_newell_3, while this approach is suitable for convex/concave facets with an arbitrary number of vertices, I think it is suboptimal for triangular facets. The function CGAL::orthogonal_vector is much better for this task, and we can easily switch between these two methods based on the number of edges/vertices.

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): Linux 64bit
* Compiler: clang++ 
* Release or debug mode: Release
* Specific flags used (if any): clang++ polyhedron_3_to_nef_3_benchmark.cpp -DCGAL_DO_NOT_USE_BOOST_MP -lgmp -lmpfr -lbenchmark
* CGAL version: 5.0.2
* Boost version: 1.71.0
* Other libraries versions if used (Eigen, TBB, etc.): Qt 5.12.8

### PR DESCRIPTION
**Pull request by GilesBathgate**
## Summary of Changes

Calculate Object_index hash tables as needed i.e when the assertion fails, or when trace message is built.

## Release Management

* Affected package(s): Nef_3, Nef_S2
* Issue(s) solved (if any): fix #5379
* Feature/Small Feature (if any): performance/speed
* License and copyright ownership: Returned to CGAL authors.

### CODE REVIEW

File: Nef_S2/include/CGAL/Nef_S2/SM_const_decorator.h [hunk 1]
```diff
@@ -273,36 +273,51 @@ The macros are then |CGAL_forall_svertices(v,M)|,
...
       ++iso_vert_num;
     } else {
-      CGAL_assertion_msg(first_out_edge(v) != SHalfedge_const_handle(),
-      VI(v).c_str());
-      CGAL_NEF_TRACEN(v->point()<<" "<<EI(first_out_edge(v)));
-      CGAL_assertion_msg(first_out_edge(v)->source() == v,
-                         VI(v).c_str());
+      CGAL_assertion_code(SHalfedge_const_handle e=first_out_edge(v));
+      CGAL_assertion_msg(e != SHalfedge_const_handle(), get_svertex_index(v).c_str());
+      CGAL_NEF_TRACEN(v->point()<<" "<<get_shalfedge_index(e));
```
Comment (GilesBathgate) [inline Nef_S2/include/CGAL/Nef_S2/SM_const_decorator.h:319]: A better approach here would be to keep EI in local scope but enclosed by `#ifdef CGAL_NEF_TRACE`, which is implemented in #5246

**Review by afabri — APPROVED** [review]
`VI`  could also be in a `CGAL_assertion_code()`, but as `EI` cannot because it is in a `CGAL_NEF_TRACEN()`  I approve your changes.

**Comment by GilesBathgate** [comment]
@afabri Yes I did consider wrapping the method `get_svertex_index` in a `CGAL_assertion_code` but with `CGAL_NO_ASERTIONS`  it would just be an unused method which is ok?

**Review by sloriot — APPROVED** [review]

**Comment by sloriot** [comment]
We are only waiting for the testsuite to be green to merge this branch.

**Comment by maxGimeno** [comment]
Success in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.3-Ic-89.shtml

**Pull request merged**